.circleci/circle-ci-config.py

```diff
@@ -0,0 +1,167 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import argparse
+import math
+import os
+import sys
+
+bin_dir = os.path.dirname(os.path.realpath(__file__))
+src_dir = bin_dir
+dep_dir = os.path.join(bin_dir, '.python3-deps')
+sys.path.append(dep_dir)
+
+try:
+    import yaml
+except ImportError:
+    import subprocess
+    # dependencies not installed; install them outside of the normal system
+
+    subprocess.check_call([sys.executable, '-m', 'pip', 'install', "--target=" + dep_dir, 'pyyaml'])
+
+    import yaml
+
+try:
+    from yaml import CLoader as Loader, CDumper as Dumper
+except ImportError:
+    from yaml import Loader, Dumper
+
+targets = {
+        'MIDRES': {
+            # apply defaults to all jobs, so the below just overrides jobs which do not match the default
+            '_default_job_':                            {'parallelism': 25,    'exec_resource_class': 'large'},
+            'jobs': {
+                'j8_jvm_upgrade_dtests':                {'parallelism': 4},
+                'j11_jvm_dtests':                       {'parallelism': 5},
+                'j11_jvm_dtests_vnode':                 {'parallelism': 5},
+                'j8_jvm_dtests':                        {'parallelism': 5},
+                'j8_jvm_dtests_vnode':                  {'parallelism': 5},
+
+                'j11_unit_tests':                       {'parallelism': 50, 'exec_resource_class': 'medium'},
+                'j11_cqlsh-dtests-py3-no-vnodes':       {'parallelism': 50, 'exec_resource_class': 'medium'},
+                'j11_cqlsh-dtests-py3-with-vnodes':     {'parallelism': 50, 'exec_resource_class': 'medium'},
+                'j11_cqlsh-dtests-py38-no-vnodes':      {'parallelism': 50, 'exec_resource_class': 'medium'},
+                'j11_cqlsh-dtests-py38-with-vnodes':    {'parallelism': 50, 'exec_resource_class': 'medium'},
+                'j11_dtests-no-vnodes':                 {'parallelism': 50},
+                'j11_dtests-with-vnodes':               {'parallelism': 50},
+
+                'repeated_upgrade_dtest':               {'exec_resource_class': 'xlarge'},
+            },
+        },
+        'HIGHER': {
+            # apply defaults to all jobs, so the below just overrides jobs which do not match the default
+            '_default_job_':                            {'parallelism': 100,    'exec_resource_class': 'xlarge'},
+            'jobs': {
+                'j8_jvm_upgrade_dtests':                {'parallelism': 2},
+                'j11_jvm_dtests':                       {'parallelism': 5},
+                'j11_jvm_dtests_vnode':                 {'parallelism': 5},
+                'j8_jvm_dtests':                        {'parallelism': 5},
+                'j8_jvm_dtests_vnode':                  {'parallelism': 5},
+            },
+        },
+}
```

Comment:
The core idea is that we maintain the list of jobs which need different resources, rather than patches which assume offers do not change...

This also doesn't rely on custom executors as executors are created per job already, so rather than having many executors, we can really get away with 2-4 executors (j8/j11, java/python) and define resources per job